File: backend/GameApp.Domain/ValueObjects/Scenes/SceneDescription.cs
namespace GameApp.Domain.ValueObjects.Scenes;


// Value Object for a game's scene description
public class SceneDescription
{
    private static readonly string _messageIfEmpty = "Scene description cannot be empty";
    public string Description { get; private set; } = default!;


    public SceneDescription(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new ArgumentException(_messageIfEmpty, nameof(value));

        Description = value;
    }

    public string GetDescription() => Description;

    public SceneDescription SetDescription(string newDescription)
    {
        if (string.IsNullOrWhiteSpace(newDescription))
            throw new ArgumentException(_messageIfEmpty, nameof(newDescription));

        return new SceneDescription(newDescription);
    }

    public override bool Equals(object? obj)
    {
        if (obj is not SceneDescription other)
            return false;

        return Description == other.Description;
    }

    public override int GetHashCode() => Description.GetHashCode();

    public override string ToString() => Description;
}

File: frontend/gameapp-frontend/src/domain/value-objects/scenes/scene-description.js
// Value Object for Scene Description
export class SceneDescription {
    static ERROR_EMPTY = "Scene description cannot be empty";

    constructor(description) {
        this.#validateDescription(description);
        this._description = description;
    }

    #validateDescription(desc) {
        if (!desc || desc.trim().length === 0) throw new Error(SceneDescription.ERROR_EMPTY);
    }

    // getter
    get description() { return this._description; }

    // setter
    setDescription(newDescription) {
        this.#validateDescription(newDescription);
        return new SceneDescription(newDescription);
    }

    equals(other) {
        return other instanceof SceneDescription && this._description === other._description;
    }

    toString() { return this._description; }
}